# Tugas Pengenalan Pola Citra

Almas Fauzia Wibawa (17/409427/PA/17734)

 ----------------------------------------------------------------------------------------------------------------------

Dalam mengerjakan tugas ini, saya akan menggunakan 5 citra wajah. Kelima citra tersebut akan melalui proses preprocessing, ekstrak fitur warna, dan perhitungan similaritas.

Sebagai gambaran, citra wajah yang saya gunakan saya beri nama file dari 0 sampai 4. Kesemuanya adalah citra aktor dan aktris Hollywood, secara berurutan yaitu: (1) Chris Hemsworth, (2) Liam Hemsworth, (3) Emma Watson, (4) Lupita Nyong'o, dan (5) Idris Elba. Alasan dipilihnya kelima orang itu adalah:

1.   Chris Hemsworth = bersaudara dengan Liam Hemsworth (citra kedua) dan mewakili individu laki-laki berkulit putih.
2.   Liam Hemsworth = bersaudara dengan Chris Hemsworth (citra pertama) dan mewakili individu laki-laki berkulit putih.
3.   Emma Watson = mewakili individu perempuan berkulit putih.
4.   Lupita Nyong'o = mewakili individu perempuan berkulit hitam.
5.   Idris Elba = mewakili individu laki-laki berkulit hitam.

Terdapat beberapa kesamaan yang dimiliki kelima individu tersebut. Diharapkan, dapat dilihat kesamaan yang mana yang ditangkap dengan membandingkan fitur warna pada kelima citra wajah mereka.

### 1. Import Library

Pertama dan yang utama, saya perlu terlebih dahulu meng-import library yang akan saya gunakan di sepanjang proses preprocessing, ekstraksi fitur, dan menghitung similaritas.

Saya menggunakan library Numpy untuk mengolah piksel citra dengan array, CV2 untuk mengolah citranya, dan pandas untuk menampilkan tabel perbandingan nilai similaritasnya.

In [136]:
import numpy as np
import cv2
import pandas

### 2. Preprocessing

Yang selanjutnya saya lakukan adalah preprocessing. Karena kelima citra yang saya gunakan sama jenisnya, yaitu wajah, preprocessing yang dilakukan kepada kelima citra tersebut pun juga akan sama. Oleh karena itu, saya akan menggunakan perulangan untuk preprocessing-nya.

Pada proses preprocessing ini, pertama-tama, saya akan meng-import citra yang akan diolah. Kemudian, saya akan lakukan deteksi wajah. Selanjutnya, berdasarkan deteksi wajah tersebut, saya akan potong citra agar hanya wajah yang ditampilkan. Saya juga perlu menyamakan ukuran citra setelah preprocessing. Oleh karena itu, saya akan menyimpan ukuran terkecil citra hasil pemotongan di dalam variabel minSize. Kelima citra hasil pemotongan akan kemudian saya resize dengan ukuran citra terkecil tersebut.

Saya tidak melakukan perubahan warna ke grayscale karena fitur yang akan diekstrak adalah fitur warna. Dengan begitu, perubahan citra ke grayscale menjadi hal yang tidak perlu dilakukan.

In [137]:
minSize = 0
images = []
n = 5

for i in range(n):
    #import citra
    image = cv2.imread(str(i) + '.jpg')
    imageRGB = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    #deteksi wajah
    face_cascade = cv2.CascadeClassifier(
        'C:/ProgramData/Anaconda3/Lib/site-packages/cv2/data/haarcascade_frontalface_default.xml')
    eye_cascade = cv2.CascadeClassifier('C:/ProgramData/Anaconda3/Lib/site-packages/cv2/data/haarcascade_eye.xml')
    faces = face_cascade.detectMultiScale(imageRGB, 1.25, 5)

    print('Number of faces detected:', len(faces))

    #memotong citra
    for f in faces:
        x, y, w, h = [ v for v in f ]
        face = imageRGB[y:y+h, x:x+w]

    #menyimpan size terkecil dari semua citra
    if (i == 0) | (face.shape[0] < minSize):
        minSize = face.shape[0]
    
    #menyimpan sementara piksel citra
    images.append(face)

#resize citra
for i in range(n):
    images[i] = cv2.resize(images[i], (minSize, minSize))
    cv2.imwrite('cropped' + str(i) + '.jpg',cv2.cvtColor(images[i], cv2.COLOR_RGB2BGR))

Number of faces detected: 1
Number of faces detected: 1
Number of faces detected: 1
Number of faces detected: 1
Number of faces detected: 1


### 3. Ekstraksi Fitur Warna

Setelah dilakukan preprocessing, dilakukan ekstraksi fitur. Fitur yang akan diekstrak kali ini adalah fitur warna pada piksel. Fitur dari setiap citra disimpan masing-masing dalam satu array numpy, yang kemudian disatukan dalam satu array. Piksel disimpan dalam bentuk RGB.

In [139]:
pixels = []
for i in range(n):
    pixels.append(np.asarray(cv2.cvtColor(cv2.imread('cropped' + str(i) + '.jpg'), cv2.COLOR_BGR2RGB)))

### 4. Menghitung Similaritas

Saya sudah mendapatkan fitur warna dari kelima citra hasil preprocessing. Yang selanjutnya dilakukan adalah menghitung similaritas antara kelima citra tersebut.

Perhitungan similaritas yang saya gunakan adalah perhitungan jarak dengan metode Euclidean Distance. Setelah ditemukan similaritasnya, akan langsung saya tampilkan tabel perbandingannya.

In [140]:
#fungsi perhitungan Euclidean Distance
def euclidean(img1, img2):
    if (img1.shape != img2.shape):
        return False
    else:
        distance = 0
        n = img1.shape[0]
        m = img1.shape[1]
        
        for i in range(n):
            for j in range(m):
                distance = np.sum((img1[i][j]-img2[i][j])**2)**0.5
        return distance

In [141]:
similarity = [[None for i in range(n)] for j in range(n)]

#menghitung similaritas antar citra
for i in range(n):
    for j in range(n):
        similarity[i][j] = euclidean(pixels[i], pixels[j])

#menampilkan tabel perbandingan similaritas
pandas.DataFrame(similarity)

0          1          2          3          4
0   0.000000  21.023796  12.806248  14.966630  16.613248
1  21.023796   0.000000  23.194827  12.727922  14.628739
2  12.806248  23.194827   0.000000  15.620499  16.733201
3  14.966630  12.727922  15.620499   0.000000  15.099669
4  16.613248  14.628739  16.733201  15.099669   0.000000

Tabel di atas adalah tabel perbandingan similaritas citra wajah kelima individu yang telah saya jelaskan di awal.

Jika diurutkan berdasarkan nilai similaritas citra wajahnya, pasangan aktor/aktris yang dinilai mirip adalah sebagai berikut:

1. (12.727922) Lupita Nyong'o - Idris Elba
2. (12.806248) Chris Hemsworth - Liam Hemsworth
3. (14.628739) Emma Watson - Lupita Nyong'o
4. (14.966630) Chris Hemsworth - Idris Elba
5. (15.099669) Emma Watson - Idris Elba
6. (15.620499) Liam Hemsworth - Idris Elba
7. (16.613248) Chris Hemsworth - Emma Watson
8. (16.733201) Liam Hemsworth - Emma Watson
9. (21.023796) Chris Hemsworth - Lupita Nyong'o
10. (23.194827) Liam Hemsworth - Lupita Nyong'o

Urutan pertama, kedua, dan ketiga merupakan salah satu petunjuk bahwa warna kulit hitam, hubungan saudara, dan jenis kelamin perempuan dapat dinilai kesamaannya dengan metode Euclidean Distance dengan fitur warna. Sedangkan kesamaan jenis kelamin laki-laki dan warna kulit putih kurang dapat dinilai kesamaannya dengan metode ini karena kesamaan terhadap keduanya berada di peringkat 4, 6, 7, dan 8. Pasangan Chris Hemsworth dengan Lupita Nyong'o dan Liam Hemsworth dengan Lupita Nyong'o terasa sangat tepat berada di peringkat 9 dan 10 dengan nilai yang terpaut agak jauh dengan peringkat atasnya dikarenakan tidak adanya kesamaan di antara kedua individu dari dua pasangan tersebut.

Walaupun begitu, metode ini dapat dinilai kurang efektif digunakan untuk pengenalan pola karena perbedaan antar pasangannya tidak cukup jauh sehingga nilai error kemungkinan masih ada. Selain itu, bukti lain yang menunjukkan bahwa metode ini kurang tepat adalah bertempatnya pasangan Emma Watson dan Idris Elba, yang tidak memiliki kesamaan wajah apapun, berada diperingkat 5. Peringkat yang cukup tinggi dengan nilai kesamaan yang tidak jauh berbeda dengan peringkat atasnya.

#  

###### Disklaimer!

Penulis tidak bermaksud rasis atau seksis. Penggolongan berdasarkan warna kulit dan jenis kelamin hanya dilakukan dengan tujuan analisa salah satu metode pengenalan pola berdasarkan fitur warna piksel pada citra wajah.